File: backend/app/services/llm_routing.py
"""Which engine serves each unit of work, derived from the code that decides it.

Luminary's claim is that a library stays on the machine and only a question ever
leaves. That is only checkable if something reports where each unit of work
actually runs, so this builds the report the Settings surface renders.

**Nothing here asserts an engine.** The routable roles resolve through
`model_router.resolve`, which is what `LLMService` itself calls, and the fixed
components report the id held by the module that loads them. `on_device` is then
read off that id against the provider set routing uses to build a model string
(`settings_service.cloud_provider_prefixes`), so a component that ever acquires a
hosted implementation flips this table without anyone editing it. A hand-written
column would state the claim rather than measure it, and would keep stating it
after a route moved -- which is the failure this module exists to make impossible.

`resolve` never raises, so a cloud route missing its key appears here as the local
model it will really use, carrying the reason. That case is the one worth seeing:
the mode says cloud and the machine is answering.
"""

from __future__ import annotations

from dataclasses import dataclass, field

from app.config import get_settings
from app.services import settings_service
from app.services.embedder import MODEL_NAME as EMBEDDING_MODEL
from app.services.model_router import resolve


@dataclass(frozen=True)
class WorkRouting:
    id: str
    label: str
    # None only where no model is involved at all. A null model and an unresolved
    # one are different answers, so nothing substitutes a placeholder here.
    model: str | None
    on_device: bool
    # False where the mode cannot move this work at any setting. The distinction
    # is the point of the table: four rows are local by choice, five by
    # construction, and collapsing them would let a reader think the whole
    # pipeline follows the mode.
    routable: bool
    why: str
    # Set when resolution fell back -- a missing key, a host that cannot hold the
    # configured model. Reported, never swallowed.
    fallback_reason: str | None = None


@dataclass(frozen=True)
class RoutingReport:
    mode: str
    provider: str | None
    work: list[WorkRouting] = field(default_factory=list)

    @property
    def leaves_device(self) -> list[str]:
        return [w.id for w in self.work if not w.on_device]


def _on_device(model_id: str | None) -> bool:
    """Whether *model_id* names something that runs on this machine.

    Off-device is decided by the provider prefix routing itself prepends, not by
    a list kept here: `get_effective_routing` builds `"{provider}/{model}"` from
    the same set. A local id is either bare, an `ollama/` id, or a HuggingFace
    `org/name` -- none of which can collide with a provider, because a provider
    prefix is exactly what makes LiteLLM send the call somewhere.
    """
    if model_id is None:
        return True
    prefix, _, rest = model_id.partition("/")
    if not rest:
        return True
    return prefix not in settings_service.cloud_provider_prefixes()


def _routed(work_id: str, label: str, role: str, why: str) -> WorkRouting:
    choice = resolve(role)  # type: ignore[arg-type]
    return WorkRouting(
        id=work_id,
        label=label,
        model=choice.model,
        on_device=_on_device(choice.model),
        routable=True,
        why=why,
        fallback_reason=choice.fallback_reason,
    )


def _fixed(work_id: str, label: str, model: str | None, why: str) -> WorkRouting:
    return WorkRouting(
        id=work_id,
        label=label,
        model=model,
        on_device=_on_device(model),
        routable=False,
        why=why,
    )


def routing_report() -> RoutingReport:
    """Every unit of work, in pipeline order, with the engine that serves it.

    Answering sits last among the model rows because in hybrid mode it is the
    one that leaves, and a table read top to bottom should arrive there.
    """
    settings = get_settings()
    mode, provider = settings_service.active_routing()

    work = [
        _fixed(
            "indexing",
            "Indexing your library",
            EMBEDDING_MODEL,
            "One embedding space holds every stored vector, so this cannot be "
            "routed without regenerating all of them.",
        ),
        _fixed(
            "retrieval",
            "Finding the passage",
            settings.RERANK_MODEL,
            "Search runs against the local index; the cross-encoder re-scores what it returns.",
        ),
        _fixed(
            "transcription",
            "Transcribing audio and video",
            f"whisper-{settings.WHISPER_MODEL_SIZE}",
            "Whisper runs on this machine in every mode.",
        ),
        _fixed(
            "extraction",
            "Extracting topics and entities",
            settings.NER_MODEL,
            "The extractor is loaded in this process; there is no remote path.",
        ),
        _routed(
            "enrichment",
            "Summaries, tags and titles at ingest",
            "background",
            "Background work stays on the machine in hybrid mode, so ingesting a "
            "library never spends your quota.",
        ),
        _routed(
            "figures",
            "Reading figures",
            "vision",
            "The figure reader is an on-device model chosen for this host.",
        ),
        _routed(
            "study_material",
            "Writing cards and study material",
            "generation",
            "You asked for it, so it follows the interactive route.",
        ),
        _routed(
            "answering",
            "Answering your question",
            "chat",
            "Your question and the passages retrieved for it are what the "
            "answering model receives.",
        ),
        _fixed(
            "learner_record",
            "Your learner record",
            None,
            "Scheduling, mastery and calibration are computed from your own "
            "review events. No model is involved, so there is nothing to route.",
        ),
    ]

    return RoutingReport(mode=mode, provider=provider, work=work)

File: frontend/src/lib/llmRouting.ts
// Reading GET /settings/llm/routing, and the one sentence that summarises it.
//
// Separate from the component so the summary can be unit-tested: Vitest runs in
// the node environment here, so pure logic lives in lib/ and components stay
// untested by convention.

import { apiGet } from "@/lib/apiClient"
import type { components } from "@/types/api"

export type RoutingResponse = components["schemas"]["RoutingResponse"]
export type WorkRoutingItem = components["schemas"]["WorkRoutingItem"]

export function fetchRouting(): Promise<RoutingResponse> {
  return apiGet<RoutingResponse>("/settings/llm/routing")
}

/**
 * The sentence under the routing table.
 *
 * Derived from the rows rather than written per mode. Three hardcoded sentences
 * used to sit here, and one of them was wrong for a year -- it told the reader
 * flashcards stayed local in hybrid mode while card generation took the
 * interactive route. A summary computed from the rows cannot drift from them.
 */
export function egressSummary(data: RoutingResponse): string {
  const notRun = data.work.filter((w) => w.refused_reason)
  const notRunLine =
    notRun.length > 0
      ? ` Not run on this machine: ${joinLabels(notRun.map((w) => w.label.toLowerCase()))}.`
      : ""
  const leaving = data.work.filter((w) => !w.on_device)
  if (leaving.length === 0) {
    return `Nothing leaves this machine.${notRunLine}`
  }
  const provider = data.provider ?? "your provider"
  // Enrichment is the one routable row that sends document text rather than a
  // question: in Cloud mode a summary is written from a section, not a retrieval.
  const sendsSections = leaving.some((w) => w.id === "enrichment")
  const withQuestions = leaving.filter((w) => w.id !== "enrichment")
  const parts = [`Only ${joinLabels(leaving.map((w) => w.label.toLowerCase()))} reach ${provider}.`]
  if (withQuestions.length > 0) {
    parts.push(
      `${capitalise(joinLabels(withQuestions.map((w) => w.label.toLowerCase())))} send your question plus the passages retrieved for it.`,
    )
  }
  if (sendsSections) {
    parts.push("Summaries, tags and titles send sections of your documents.")
  }
  parts.push("Your library, its index and your learner record stay on this machine.")
  return `${parts.join(" ")}${notRunLine}`
}

export function joinLabels(labels: string[]): string {
  if (labels.length <= 1) return labels[0] ?? ""
  return `${labels.slice(0, -1).join(", ")} and ${labels[labels.length - 1]}`
}

function capitalise(text: string): string {
  return text.charAt(0).toUpperCase() + text.slice(1)
}
